File: scala-examples/src/_03_extensions/extensions.scala
package _03_extensions

class C


class RichC(c : C) {
  def foo() {
    println(c)
  }
}


object Main extends App {
  val c = new C()
  c.foo()

  implicit def enrich(c : C) : RichC = new RichC(c)
}










object Test2 extends App {
  val c = new C()

  takeRich(c)

  def takeRich(r : RichBase) {}
  implicit def enrich(c : C) : RichC2 = new RichC2()
}

class RichBase {
  def bar() {}
}

class RichC2 extends RichBase {
  def foo() {}
}




object TestFor extends App {

  val c = new C()
  for (i <- c) {
    println(i)
  }

  class RichC {
    def foreach(x : Int => Unit) {
      x(1)
      x(2)
    }
  }

  implicit def enrich(c : C) : RichC = new RichC()


}

File: kotlin-to-scala-examples/src/_04_extensions/extensions.kt
package extensions

class C

fun main(args : Array<String>) {
    val c = C()
    c.foo()

    c.iterator()
    for (i in c) {

    }
}










fun C.foo() {
    println(this)
}







fun test2() {
    val c = C()
    c.enrich().bar()
    takeRich(c.enrich())
}

fun takeRich(rich : RichBase) {

}

open class RichBase {
    fun bar() {}
}

class RichC(c : C) : RichBase() {

}

fun C.enrich() = RichC(this)











fun test3() {
    val c = C()
    for (i in c) {
        println(i)
    }
}

fun C.iterator() = arrayList(1, 2).iterator()